flang/lib/Lower/IO.cpp

```diff
@@ -113,29 +115,29 @@ static mlir::FuncOp getIORuntimeFunc(Fortran::lower::FirOpBuilder &builder) {
   return func;
 }
 
-/// Generate a call to end an IO statement
+/// Generate calls to end an IO statement.  Return the IOSTAT value, if any.
+/// It is the caller's responsibility to generate branches on that value.
 static mlir::Value genEndIO(Fortran::lower::AbstractConverter &converter,
                             mlir::Location loc, mlir::Value cookie,
-                            const ErrorHandling &eh) {
-  // Terminate IO
+                            const ConditionSpecifierInfo &csi) {
   auto &builder = converter.getFirOpBuilder();
-  if (eh.ioMsg) {
-    TODO(); // eh.ioMsgExpr is never set
-    auto getMsg = getIORuntimeFunc<mkIOKey(GetIoMsg)>(builder);
-    auto var = converter.genExprAddr(eh.ioMsgExpr, loc);
-    mlir::Value varLen{}; // FIXME
-    llvm::SmallVector<mlir::Value, 3> args{cookie, var, varLen};
-    builder.create<mlir::CallOp>(loc, getMsg, args);
+  if (csi.ioMsgExpr) {
+    auto getIoMsg = getIORuntimeFunc<mkIOKey(GetIoMsg)>(builder);
+    auto ioMsgVar = converter.genExprAddr(csi.ioMsgExpr, loc);
+    auto ioMsgVarLen = builder.createUnboxChar(ioMsgVar).second;
+    llvm::SmallVector<mlir::Value, 3> args{cookie, ioMsgVar, ioMsgVarLen};
```

Comment:
I think you should use:
```
    auto ioMsgUnboxed = builder.createUnboxChar(ioMsgVar);
    llvm::SmallVector<mlir::Value, 3> args{cookie, ioMsgUnboxed.first, ioMsgUnboxed.second};
```
because `ioMsgVar` may have a `fir.boxchar` type (if for instance `m` was a assumed length dummy in your example).